### PR TITLE
Port conversations backbone, use Typeahead pt. 2 and Bootstrap tag

### PR DESCRIPTION
**Pull request by ghost**
- [x] Fix cukes and Jasmine
- [x] Write tests
- [x] Fix cucumber

screen shot 2016-07-14 at 01 07 01

### CODE REVIEW

File: app/assets/javascripts/app/views/conversations_view.js [hunk 1]
```diff
@@ -1,22 +1,109 @@
 // @license magnet:?xt=urn:btih:0b31508aeb0634b347b8270c7bee4d411b5d4109&dn=agpl-3.0.txt AGPL-v3-or-Later
 
 app.views.Conversations = Backbone.View.extend({
-
   el: "#conversations_container",
 
   events: {
-    "keydown textarea#message_text" : "keyDown",
-    "conversation:loaded" : "setupConversation"
+    "keydown textarea#message-text" : "keyDown",
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_view.js]: Extra space after key 'keydown textarea#message-text'.

File: app/assets/javascripts/app/views/conversations_view.js [hunk 2]
```diff
@@ -1,22 +1,109 @@
...
+      this.search.ignorePersonForSuggestions(person);
+    }.bind(this));
+
+    this.contactsTags.getTags().length = 0;
+
+    this.conversationParticipants.forEach(function(person) {
+      this.contactsTags.addTag(person.name);
+    }.bind(this));
+
+    console.log(this.conversationParticipants);
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_view.js]: Unexpected console statement.

File: app/assets/javascripts/app/views/conversations_view.js [hunk 3]
```diff
@@ -1,22 +1,109 @@
...
+  getContactsIdsListInput: function() {
+    return this.$el.find("#as-values-contact-ids");
+  },
+
+  getContactsTagsInput: function() {
+    return this.$el.find("#contacts-tags-list");
+  },
+
+  renderConversation: function(conversationId) {
+    if(conversationId){
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_view.js]: Expected space(s) after "if".

Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_view.js]: Missing space before opening brace.

File: app/assets/javascripts/app/views/conversations_view.js [hunk 4]
```diff
@@ -1,22 +1,109 @@
...
+          self.$el.find("#conversation-inbox").removeClass("selected");
+          self.$el.find(".stream_element[data-guid='" + conversationId + "'] #conversation-inbox").addClass("selected");
+          self.setupConversation();
+        }
       });
     }
-    this.setupConversation();
+  },
+
+  showNewConversation: function(evt){
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_view.js]: Missing space before opening brace.

File: app/assets/javascripts/app/views/conversations_view.js [hunk 5]
```diff
@@ -49,10 +136,26 @@ app.views.Conversations = Backbone.View.extend({
     }
   },
 
+  displayConversation: function(evt) {
+    var $target = $(evt.target);
+    if(!$target.hasClass(".conversation-wrapper")){
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_view.js]: Expected space(s) after "if".

Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_view.js]: Missing space before opening brace.

File: app/assets/javascripts/app/views/conversations_view.js [hunk 6]
```diff
@@ -49,10 +136,26 @@ app.views.Conversations = Backbone.View.extend({
...
+
   keyDown : function(evt) {
     if(evt.which === Keycodes.ENTER && evt.ctrlKey) {
       $(evt.target).parents("form").submit();
     }
+  },
+
+  onSubmitForm: function(evt) {
+    evt.preventDefault();
+    if(this.conversationParticipants.length === 0){
```
Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_view.js]: Expected space(s) after "if".

Comment (diaspora-code-review) [inline app/assets/javascripts/app/views/conversations_view.js]: Missing space before opening brace.

File: app/views/conversations/_messages.haml [hunk 7]
```diff
@@ -8,15 +8,11 @@
       .media-body
         = form_for [conversation, Message.new], html: {class: "control-group"} do |message|
           .form-group
-            %label#messageLabel.sr-only{for: "message_text"}
+            %label#message-label.sr-only{for: "message_text"}
```
Comment (diaspora-code-review) [inline app/views/conversations/_messages.haml]: Classes should be listed before IDs (.sr-only should precede #message-label)

File: app/views/conversations/_messages.haml [hunk 8]
```diff
@@ -8,15 +8,11 @@
...
-                                tabindex: 1,
-                                class:    "form-control form-group",
-                                aria:     {labelledby: "messageLabel"}
+            = message.text_area :text, rows: 5, tabindex: 1, id: "message-text",
+                class: "form-control form-group", aria: {labelledby: "message-label"}
 
-          = message.submit t("conversations.show.reply"),
-                            "data-disable-with" => t("conversations.show.replying"),
-                            class: "btn btn-primary pull-right", tabindex: 2
+          = message.submit t("conversations.show.reply"), "data-disable-with" => t("conversations.show.replying"),
```
Comment (diaspora-code-review) [inline app/views/conversations/_messages.haml]: Don't mix styles in the same hash.

File: app/views/conversations/_new.haml [hunk 9]
```diff
@@ -1,22 +1,23 @@
 .container-fluid
-  = form_for Conversation.new, html: {class: "form-horizontal form_do_not_clear"}, remote: true do |conversation|
+  = form_for Conversation.new, html: {id: "new-conversation",
+    class: "new-conversation form-horizontal form-do-not-clear"}, remote: true do |conversation|
     .form-group
-      %label#toLabel{for: "contact_ids"}
-        = t(".to")
-      = text_field_tag "contact_autocomplete", nil, class: "form-control"
+      %label#to-label{for: "contacts-search-input"}= t(".to")
+      .clearfix#contacts-tags-list.contacts-tags-list
```
Comment (diaspora-code-review) [inline app/views/conversations/_new.haml]: Classes should be listed before IDs (.contacts-tags-list should precede #contacts-tags-list)

File: app/views/conversations/_new.haml [hunk 10]
```diff
@@ -1,22 +1,23 @@
...
+      = text_field_tag "contacts-autocomplete", nil, id: "contacts-search-input", class: "form-control"
+      - unless defined? mobile && mobile
+        = text_field_tag "contacts-ids", nil, id: "as-values-contact-ids", type: "hidden", aria: {labelledby: "to-label"}
```
Comment (diaspora-code-review) [inline app/views/conversations/_new.haml]: Line is too long. [121/120]

File: app/views/conversations/_new.haml [hunk 11]
```diff
@@ -1,22 +1,23 @@
...
+
     .form-group
-      %label#messageLabel.sr-only{for: "conversation_text"}
-        = t(".message")
-      = text_area_tag "conversation[text]",
-                      "",
-                      rows:  5,
-                      class: "input-block-level form-control",
-                      aria:  {labelledby: "messageLabel"}
+      %label#message-label.sr-only{for: "message-text"}= t(".message")
```
Comment (diaspora-code-review) [inline app/views/conversations/_new.haml]: Classes should be listed before IDs (.sr-only should precede #message-label)

File: app/controllers/conversations_controller.rb [hunk 12]
```diff
@@ -30,12 +30,22 @@ def index
...
-      person_ids = current_user.contacts.where(id: contact_ids).pluck(:person_id)
+    contacts_ids = params["contacts-ids"] || params["contacts-ids-mobile"]
+
+    # Contacts autocomplete does not work the same way on mobile and desktop
+    # Mobile returns contact ids array while desktop returns person id
+    # This will have to be removed when mobile autocomplete is ported to Typeahead
+    if defined? params["contacts-ids-mobile"]
+      # Can't split nil
+      if contacts_ids
+        contacts_ids = contacts_ids.split(',') if contacts_ids.is_a? String
```
Comment (diaspora-code-review) [inline app/controllers/conversations_controller.rb]: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

File: app/controllers/conversations_controller.rb [hunk 13]
```diff
@@ -30,12 +30,22 @@ def index
...
+        person_ids = current_user.contacts.where(id: contacts_ids).pluck(:person_id)
+      end
+    else
+      if contacts_ids
```
Comment (diaspora-code-review) [inline app/controllers/conversations_controller.rb]: Convert `if` nested inside `else` to `elsif`.

File: app/controllers/conversations_controller.rb [hunk 14]
```diff
@@ -30,12 +30,22 @@ def index
...
+        contacts_ids = contacts_ids.split(',') if contacts_ids.is_a? String
```
Comment (diaspora-code-review) [inline app/controllers/conversations_controller.rb]: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

File: app/controllers/conversations_controller.rb [hunk 15]
```diff
@@ -63,16 +73,18 @@ def create
 
   def show
     respond_to do |format|
-      format.html do
-        redirect_to conversations_path(:conversation_id => params[:id])
-        return
+      unless params[:raw]
+        format.html do
+          redirect_to conversations_path(:conversation_id => params[:id])
```
Comment (diaspora-code-review) [inline app/controllers/conversations_controller.rb]: Use the new Ruby 1.9 hash syntax.

File: app/controllers/conversations_controller.rb [hunk 16]
```diff
@@ -63,16 +73,18 @@ def create
...
+          return
+        end
       end
 
       if @conversation = current_user.conversations.where(id: params[:id]).first
         @first_unread_message_id = @conversation.first_unread_message(current_user).try(:id)
         @conversation.set_read(current_user)
 
-        format.js
+        format.html { render partial: "show", locals: { conversation: @conversation }}
```
Comment (diaspora-code-review) [inline app/controllers/conversations_controller.rb]: Space inside { detected.

Comment (diaspora-code-review) [inline app/controllers/conversations_controller.rb]: Space inside } detected.

Comment (diaspora-code-review) [inline app/controllers/conversations_controller.rb]: Space missing inside }.

File: app/assets/stylesheets/_application.scss [hunk 17]
```diff
@@ -106,6 +106,9 @@
 @import 'blueimp-gallery/blueimp-gallery-indicator';
 @import 'gallery';
 
+// Bootstrap tags
+@import "bootstrap-tags";
```
Comment (diaspora-code-review) [inline app/assets/stylesheets/_application.scss]: Prefer single quoted strings

File: app/assets/stylesheets/conversations.scss [hunk 18]
```diff
@@ -183,8 +183,18 @@
...
 // scss-lint:enable SelectorDepth
 
-#new_conversation_pane {
-  ul.as-selections { width: 100% !important; }
-  input#contact_ids { box-shadow: none; }
-  label { font-weight: bold; }
+.new-conversation {
+  .contacts-tags-list {
+    margin-bottom: $form-group-margin-bottom;
+    padding: 0 !important;
```
Comment (diaspora-code-review) [inline app/assets/stylesheets/conversations.scss]: !important should not be used

File: app/assets/stylesheets/mobile/conversations.scss [hunk 19]
```diff
@@ -61,3 +61,5 @@
 .subject { padding: 0 10px; }
 
 .message-count, .unread-message-count { margin: 10px 2px; }
+
+.new-conversation ul.as-selections { background-color: transparent; }
```
Comment (diaspora-code-review) [inline app/assets/stylesheets/mobile/conversations.scss]: Avoid qualifying class selectors with an element.

File: app/assets/stylesheets/mobile/mobile.scss [hunk 20]
```diff
@@ -597,7 +597,7 @@ form#new_user.new_user input.btn {
   text-shadow: 1px 1px 20px rgb(126, 240, 77);
 }
 
-#conversation_inbox, .notifications {
+#conversation-inbox, .notifications {
```
Comment (diaspora-code-review) [inline app/assets/stylesheets/mobile/mobile.scss]: Avoid using id selectors

Comment (diaspora-code-review) [inline app/assets/stylesheets/mobile/mobile.scss]: Each selector in a comma sequence should be on its own single line

**Comment by AugierLe42e** [comment]
Oh sh... Forgot Pronto. Sorry for spamming.

File: app/views/conversations/_messages.haml (hunk 7 above)
Comment (Flaburgan) [inline app/views/conversations/_messages.haml]: I find this rule weird, is that the default or did we force it?

**Comment by AugierLe42e** [comment]
This is in production on diaspote.org since tonight. I still have one little test to fix but this PR can be revied now.

**Comment by AugierLe42e** [comment]
I'm done. Tests are passing. Ready for review.

File: app/controllers/conversations_controller.rb [hunk 21]
```diff
@@ -21,21 +21,24 @@ def index
...
   end
 
   def create
-    contact_ids = params[:contact_ids]
-
-    # Can't split nil
-    if contact_ids
-      contact_ids = contact_ids.split(',') if contact_ids.is_a? String
-      person_ids = current_user.contacts.where(id: contact_ids).pluck(:person_id)
+    contacts_ids = params["contacts-ids"]
```
Comment (jhass) [inline app/controllers/conversations_controller.rb]: I think the parameter name should stay `contact_ids`, the double plural sounds weird.

File: app/controllers/conversations_controller.rb [hunk 22]
```diff
@@ -21,21 +21,24 @@ def index
...
+
+    # Contacts autocomplete does not work the same way on mobile and desktop
+    # Mobile returns contact ids array while desktop returns person id
+    # This will have to be removed when mobile autocomplete is ported to Typeahead
+    if session[:mobile_view] && contacts_ids
+      contacts_ids = contacts_ids.split(",") if contacts_ids.is_a? String
```
Comment (jhass) [inline app/controllers/conversations_controller.rb]: What is it if not a string? `nil`? If so, just do `if contact_ids`.

File: app/controllers/conversations_controller.rb [hunk 23]
```diff
@@ -21,21 +21,24 @@ def index
...
+      person_ids = current_user.contacts.where(id: contacts_ids).pluck(:person_id)
+    elsif contacts_ids
+      contacts_ids = contacts_ids.split(",") if contacts_ids.is_a? String
+      person_ids = current_user.contacts.where(person_id: contacts_ids).pluck(:person_id)
```
Comment (jhass) [inline app/controllers/conversations_controller.rb]: This seems to be the same in both branches, just extract it to behind the if and guard it with another `if contact_ids`.

File: app/controllers/conversations_controller.rb [hunk 24]
```diff
@@ -63,16 +66,18 @@ def create
 
   def show
     respond_to do |format|
-      format.html do
-        redirect_to conversations_path(:conversation_id => params[:id])
-        return
+      unless params[:raw]
```
Comment (jhass) [inline app/controllers/conversations_controller.rb]: This feels like it misuses this route/action as two, perhaps better add a new route.

File: app/controllers/conversations_controller.rb [hunk 25]
```diff
@@ -86,15 +91,18 @@ def new
...
     elsif params[:aspect_id]
-      @contact_ids = current_user.aspects.find(params[:aspect_id]).contacts.map{|c| c.id}.join(',')
+      gon.push conversation_prefill: current_user.aspects.find(params[:aspect_id]).contacts.map {|c| c.person.as_json }
     end
     if session[:mobile_view] == true && request.format.html?
+      @contacts_json = contacts_data.to_json
+      @contact_ids = if params[:contact_id]
+                       current_user.contacts.find(params[:contact_id]).id
+                     elsif params[:aspect_id]
+                       current_user.aspects.find(params[:aspect_id]).contacts.map(&:id).join(",")
```
Comment (jhass) [inline app/controllers/conversations_controller.rb]: `pluck(:id)`, not `map(&:id)`.

File: app/views/conversations/new.html.haml [hunk 26]
```diff
@@ -1,11 +1,8 @@
...
 :javascript
   $(document).ready(function () {
-    var data = $.parseJSON( "#{escape_javascript(@contacts_json)}" );
-    new app.views.ConversationsForm({
-      contacts: data,
-      prefillName: "#{h params[:name]}",
-      prefillValue: "#{@contact_ids}"
-    });
+    new app.views.ConversationsForm({prefills: gon.conversationPrefill});
   });
```
Comment (jhass) [inline app/views/conversations/new.html.haml]: Can't we try to get rid of this entirely while at it? Shouldn't the backbone router call it?

File: config/initializers/assets.rb [hunk 27]
```diff
@@ -2,6 +2,10 @@
 #   licensed under the Affero General Public License version 3 or later.  See
 #   the COPYRIGHT file.
 
+Rails.configuration.assets.paths.reject! do |path|
+  path.include?("rails-assets-bootstrap") && !path.include?("rails-assets-bootstrap-tags")
+end
```
Comment (jhass) [inline config/initializers/assets.rb]: ? Some explanatory comment is needed at least.

File: app/controllers/conversations_controller.rb (hunk 23 above)
Comment (AugierLe42e) [inline app/controllers/conversations_controller.rb]: No, there're not the same both branches. The `where` clause is not the same.

File: app/views/conversations/new.html.haml (hunk 26 above)
Comment (AugierLe42e) [inline app/views/conversations/new.html.haml]: This is used in the modal on the profile. The modal content is transmitted asynchronously so we need this JavaScript to be here.

File: app/controllers/conversations_controller.rb (hunk 23 above)
Comment (jhass) [inline app/controllers/conversations_controller.rb]: Oh, that's weird, submitting person ids in a parameter called contact ids is quite confusing.

File: app/views/conversations/new.html.haml (hunk 26 above)
Comment (jhass) [inline app/views/conversations/new.html.haml]: Why can't we bind to a success callback of displaying the modal or something like that?

File: app/controllers/conversations_controller.rb (hunk 23 above)
Comment (AugierLe42e) [inline app/controllers/conversations_controller.rb]: Oh, you meant do something like :

```` ruby
if params["contacts-ids"]
  contacts_ids = params["contacts-ids"].split(",")
  if session[:mobile_view]
    person_ids = current_user.contacts.where(id: contacts_ids).pluck(:person_id)
  else
    person_ids = current_user.contacts.where(person_id: contacts_ids).pluck(:person_id)
  end
end
````

?

File: app/controllers/conversations_controller.rb (hunk 24 above)
Comment (AugierLe42e) [inline app/controllers/conversations_controller.rb]: What about `conversations/raw/:id`, or `conversations/:id/raw`?

Comment (jhass) [inline app/controllers/conversations_controller.rb]: The latter, so you can just add a member route to the resources call.

File: app/controllers/conversations_controller.rb (hunk 23 above)
Comment (jhass) [inline app/controllers/conversations_controller.rb]: That's a bit better I guess. Submitting person ids as "contact-ids" (why the dash all of the sudden btw? none of our other parameters does that) is still confusing though, perhaps something like:

```rb
recipients_param = %i(contact_ids person_ids).find {|param| params[param].present? }
person_ids  = current_user.contacts.where(recipients_param => params[recipients_param].split(",")).pluck(:person_id)

**Comment by AugierLe42e** [comment]
Done.

File: app/assets/javascripts/app/views/conversations_form_view.js [hunk 28]
```diff
@@ -1,52 +1,108 @@
-// @license magnet:?xt=urn:btih:0b31508aeb0634b347b8270c7bee4d411b5d4109&dn=agpl-3.0.txt AGPL-v3-or-Later
```
Comment (jhass) [inline app/assets/javascripts/app/views/conversations_form_view.js]: Actually why did you remove this?

File: app/controllers/conversations_controller.rb [hunk 29]
```diff
@@ -64,37 +62,49 @@ def create
...
     elsif params[:aspect_id]
-      @contact_ids = current_user.aspects.find(params[:aspect_id]).contacts.map{|c| c.id}.join(',')
+      gon.push conversation_prefill: current_user.aspects.find(params[:aspect_id]).contacts.map {|c| c.person.as_json }
     end
     if session[:mobile_view] == true && request.format.html?
+      @contacts_json = contacts_data.to_json
+      @contact_ids = if params[:contact_id]
+                       current_user.contacts.find(params[:contact_id]).id
+                     elsif params[:aspect_id]
+                       current_user.aspects.find(params[:aspect_id]).contacts.pluck(&:id).join(",")
```
Comment (jhass) [inline app/controllers/conversations_controller.rb]: Just `pluck(:id)`, no need to pass a block.

File: config/initializers/assets.rb [hunk 30]
```diff
@@ -2,6 +2,13 @@
 #   licensed under the Affero General Public License version 3 or later.  See
 #   the COPYRIGHT file.
 
+# bootstrap-tags plugin relies on rails-assets-bootstrap gem but we use
+# bootstrap-sass this line makes sure we exclude every asset comming
+# from rails-assets-bootstrap to prevent conflicts with bootstrap-sass
+Rails.configuration.assets.paths.reject! do |path|
+  path.include?("rails-assets-bootstrap") && !path.include?("rails-assets-bootstrap-tags")
+end
```
Comment (jhass) [inline config/initializers/assets.rb]: And a `require: false` in the `Gemfile` doesn't prevent it from getting loaded? :(

Comment (AugierLe42e) [inline config/initializers/assets.rb]: Unhopefully, no. I had the same problem on #6551. Using `require: false` was already proposed as a solution [here](https://github.com/tenex/rails-assets/issues/132) but it didn't help.

**Comment by AugierLe42e** [comment]
Just rebased this one. Ready for a JS review.

**Comment by AugierLe42e** [comment]
I refactored the code a bit so that the tag also shows the handle:

screen shot 2016-08-13 at 14 53 10

This will be handy for messages with several reciepiants that have the same pseudo.

**Comment by jhass** [comment]
Can we only show it if it differs from the display name?

**Comment by AugierLe42e** [comment]
Sure, I can do it.

**Comment by AugierLe42e** [comment]
screen shot 2016-08-13 at 15 22 50

I just need to write some more tests before pushing.

**Comment by AugierLe42e** [comment]
Ready. The test have been written.

File: spec/controllers/conversations_controller_spec.rb [hunk 31]
```diff
@@ -16,48 +16,56 @@
...
+         '"}]});alert(1);(function f() {var foo = [{b:"'].each do |xss|
+          get :new, modal: true, name: xss
+          expect(response.body).not_to include xss
+        end
+      end
     end
 
-    it "does not allow XSS via the name parameter" do
-      ["</script><script>alert(1);</script>",
-       '"}]});alert(1);(function f() {var foo = [{b:"'].each do |xss|
```
Comment (diaspora-code-review) [inline spec/controllers/conversations_controller_spec.rb]: Use the new Ruby 1.9 hash syntax.

File: spec/controllers/conversations_controller_spec.rb [hunk 32]
```diff
@@ -16,48 +16,56 @@
...
-        get :new, :modal => true, name: xss
```
Comment (diaspora-code-review) [inline spec/controllers/conversations_controller_spec.rb]: Line is too long. [122/120]

File: spec/controllers/conversations_controller_spec.rb [hunk 33]
```diff
@@ -115,207 +123,347 @@
...
-      it 'does not create a conversation' do
-        count = Conversation.count
-        post :create, @hash
-        expect(Conversation.count).to eq(count)
-      end
+      context "with empty contact" do
+        before do
+          @hash = {
+            format:       :js,
+            conversation: {subject: "secret stuff", text: "text debug"},
```
Comment (diaspora-code-review) [inline spec/controllers/conversations_controller_spec.rb]: Extra blank line detected.

Comment (AugierLe42e) [inline spec/controllers/conversations_controller_spec.rb]: These remarks were marked as *info* by Pronto which, consequently didn't fail and didn't prevent me to push to the repo. Maybe there's a setting womewhere to fiw that?

File: app/assets/javascripts/app/views/conversations_form_view.js [hunk 34]
```diff
@@ -1,52 +1,164 @@
...
-      this.displayNoContactsMessage();
-      return;
+    this.conversationParticipants = [];
+
+    this.search = new app.views.SearchBase({
+      el: this.$el.find("#new-conversation"),
+      typeaheadInput: this.getTypeaheadElement(),
+      customSearch: true,
+      autoselect: true,
+      remoteRoute: "/contacts"
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/conversations_form_view.js]: We should make sure that the route we use only returns mutual contacts.

Comment (AugierLe42e) [inline app/assets/javascripts/app/views/conversations_form_view.js]: Hmm... Right. So, should I create a new route or just use a parameter in `ContactsController#index`?

Comment (svbergerem) [inline app/assets/javascripts/app/views/conversations_form_view.js]: Adding a parameter sounds good to me.

Comment (AugierLe42e) [inline app/assets/javascripts/app/views/conversations_form_view.js]: Done.

**Comment by AugierLe42e** [comment]
Ping.

**Comment by svbergerem** [comment]
I started reviewing your code and got stuck in `app/assets/javascripts/app/views/conversations_form_view.js`. You add `\u200B` in front of handles because otherwise when the user removes a contact from the recipients list you won't be able to find out which contact has been removed. This is because [bootstrap-tags](https://github.com/maxwells/bootstrap-tags) only accepts strings as tags. I'm none too happy about adding hacks to support external libraries, especially if the last release was in 2014 and the only change afterwards was a README update. A feature request that could fix this for us [has been around for more than 2 years](https://github.com/maxwells/bootstrap-tags/issues/39).

I looked around for other libs that are actively maintained and would support adding ids to tags but I wasn't able to find any. Your current code looks like we are only using bootstrap-tags to render the tags and call a function when the user removes a tag and do everything else on our own. This is something we could do with a simple backbone view that renders a handlebars template. In this case we would be able to get the diaspora handle and the contact id from the backbone model for the view. We would even be able to add avatars to the list of recipients. (Like http://bootstrap-tags.azurewebsites.net/)

Before doing any further changes it would be great if you could open a new PR with the backbone port of the conversations view. (your first commit and some parts of the second commit) This makes it easier for me to review the code and we can concentrate later on the changes that might need some further discussion without delaying the backbone port.

**Comment by AugierLe42e** [comment]
Closing in favour of #7129.
